devices/flash-zynq7000/flashdrv.c

```diff
@@ -0,0 +1,461 @@
+{
+	int i;
+	size_t sz;
+	addr_t start = 0;
+	const flash_cfi_t *cfi = &fdrv_common.info.cfi;
+
+	for (i = 0; i < cfi->regsCount; ++i) {
+		sz = CFI_SIZE_REGION(cfi->regs[i].size, cfi->regs[i].count);
+		if (offs >= start && offs < sz)
+			return i;
+
+		start = sz;
+	}
+
+	return -EINVAL;
+}
+
+
+static int flashdrv_sectorErase(addr_t offs)
+{
+	int i;
+	ssize_t res;
+	time_t start, timeout;
+	unsigned int st;
+	const flash_cmd_t *cmd;
+	const flash_cfi_t *cfi = &fdrv_common.info.cfi;
+
+	if (offs > CFI_SIZE_FLASH(cfi->chipSize))
+		return -EINVAL;
+
+	if ((i = flashdrv_regionFind(offs)) < 0)
+		return i;
+
+	switch (CFI_SIZE_SECTION(cfi->regs[i].size)) {
+		case 0x1000:
+			cmd = &fdrv_common.info.cmds[flash_cmd_p4e];
+			break;
+		case 0x10000:
+			cmd = &fdrv_common.info.cmds[flash_cmd_se];
+			break;
+		default:
+			return -EINVAL;
+	}
+
+	if ((res = flashdrv_welSet(flash_cmd_wren)) < 0)
+		return res;
+
+	hal_memset(fdrv_common.cmdTx, 0, cmd->size);
+	fdrv_common.cmdTx[0] = cmd->opCode;
+	fdrv_common.cmdTx[1] = (offs >> 16) & 0xff;
+	fdrv_common.cmdTx[2] = (offs >> 8) & 0xff;
+	fdrv_common.cmdTx[3] = offs & 0xff;
+
+	qspi_start();
+	if ((res = qspi_polledTransfer(fdrv_common.cmdTx, fdrv_common.cmdRx, cmd->size, TIMEOUT_CMD_MS)) < 0)
+		return res;
+	qspi_stop();
+
+	timeout = CFI_TIMEOUT_MAX_ERASE(cfi->timeoutTypical.sectorErase, cfi->timeoutMax.sectorErase) + TIMEOUT_CMD_MS;
+	start = hal_timerGet();
+
+	/* Wait on sector erase command completion */
+	do {
+		if ((res = flashdrv_statusRegGet(flash_cmd_rdsr1, &st)) < 0)
+			return res;
+
+		if ((hal_timerGet() - start) >= timeout)
+			return -ETIME;
+	} while ((st >> 8) & 0x1);
+
+	return EOK;
+}
+
+
+static ssize_t flashdrv_pageProgram(addr_t offs, const void *buff, size_t len)
+{
+	ssize_t res;
+	time_t start, timeout;
+	unsigned int st;
+	const flash_cfi_t *cfi = &fdrv_common.info.cfi;
+	const flash_cmd_t *cmd = &fdrv_common.info.cmds[flash_cmd_pp];
+
+	if (len == 0)
+		return 0;
+
+	if (buff == NULL || len % CFI_SIZE_PAGE(cfi->pageSize))
+		return -EINVAL;
+
+	if ((res = flashdrv_welSet(flash_cmd_wren)) < 0)
+		return res;
+
+	hal_memset(fdrv_common.cmdTx, 0, cmd->size);
+	fdrv_common.cmdTx[0] = cmd->opCode;
+	fdrv_common.cmdTx[1] = (offs >> 16) & 0xff;
+	fdrv_common.cmdTx[2] = (offs >> 8) & 0xff;
+	fdrv_common.cmdTx[3] = offs & 0xff;
+
+	qspi_start();
+	if ((res = qspi_polledTransfer(fdrv_common.cmdTx, fdrv_common.cmdRx, cmd->size, TIMEOUT_CMD_MS)) < 0) {
+		qspi_stop();
+		return res;
+	}
+
+	/* Transmit data to write */
+	if ((res = qspi_polledTransfer(buff, NULL, len, TIMEOUT_CMD_MS)) < 0) {
+		qspi_stop();
+		return res;
+	}
+	qspi_stop();
+
+	len = res;
+
+	timeout = CFI_TIMEOUT_MAX_PROGRAM(cfi->timeoutTypical.pageWrite, cfi->timeoutMax.pageWrite) + TIMEOUT_CMD_MS;
+	start = hal_timerGet();
+	do {
+		if ((res = flashdrv_statusRegGet(flash_cmd_rdsr1, &st)) < 0)
+			return res;
+
+		if ((hal_timerGet() - start) >= timeout)
+			return -ETIME;
+	} while ((st >> 8) & 0x1);
```

Comment:
Copy-paste code, making this a function might be a good idea

Reply:
Yeah, good suggestion. Moved this code to separate function.